### PR TITLE
Give a better error when sygus grammar rules contain free variables.

### PR DESCRIPTION
**Pull request by abdoo8080**

### CODE REVIEW

**Review by ajreynol — COMMENTED** [review]
This looks good to me. I'd prefer @aniemetz also take a quick look, since she may be refactoring the checks for this class.

File: src/api/cvc4cpp.cpp [hunk 1]
```diff
@@ -3988,6 +3998,24 @@ void Grammar::addSygusConstructorVariables(DatatypeDecl& dt,
   CVC4_API_TRY_CATCH_END;
 }
 
+bool Grammar::containsFreeVariables(const Term& rule) const
+{
+  std::unordered_set<TNode, TNodeHashFunction> scope;
+
+  for (Term sygusVar : d_sygusVars)
```
Comment (aniemetz) [inline src/api/cvc4cpp.cpp:4005]:
```suggestion
  for (const Term& sygusVar : d_sygusVars)
```

File: src/api/cvc4cpp.cpp [hunk 2]
```diff
@@ -3988,6 +3998,24 @@ void Grammar::addSygusConstructorVariables(DatatypeDecl& dt,
...
+  {
+    scope.emplace(*sygusVar.d_node);
+  }
+
+  for (Term ntsymbol : d_ntSyms)
```
Comment (aniemetz) [inline src/api/cvc4cpp.cpp:4010]:
```suggestion
  for (const Term& ntsymbol : d_ntSyms)
```

File: src/api/cvc4cpp.cpp [hunk 3]
```diff
@@ -3676,6 +3680,12 @@ void Grammar::addRules(const Term& ntSymbol, const std::vector<Term>& rules)
       d_ntsToTerms.find(ntSymbol) != d_ntsToTerms.cend(), ntSymbol)
       << "ntSymbol to be one of the non-terminal symbols given in the "
          "predeclaration";
+  for (size_t i = 0, n = i < rules.size(); i < n; ++i) {
```
Comment (aniemetz) [inline src/api/cvc4cpp.cpp:3683]: Format

**Review by aniemetz — CHANGES_REQUESTED** [review]

File: src/api/cvc4cpp.cpp [hunk 4]
```diff
@@ -3659,6 +3660,9 @@ void Grammar::addRule(const Term& ntSymbol, const Term& rule)
          "predeclaration";
   CVC4_API_CHECK(ntSymbol.d_node->getType() == rule.d_node->getType())
       << "Expected ntSymbol and rule to have the same sort";
+  CVC4_API_ARG_CHECK_EXPECTED(!containsFreeVariables(rule), rule)
+      << "a constant term or a term containing synthFun/synthInv parameters "
+         "and non-terminal symbols of the grammar";
```
Comment (ajreynol) [inline src/api/cvc4cpp.cpp:3665]:
```suggestion
      << "Expected a term whose free variables are limited to synthFun/synthInv parameters "
         "and non-terminal symbols of the grammar";
```

File: src/api/cvc4cpp.cpp [hunk 5]
```diff
@@ -3676,6 +3680,13 @@ void Grammar::addRules(const Term& ntSymbol, const std::vector<Term>& rules)
       d_ntsToTerms.find(ntSymbol) != d_ntsToTerms.cend(), ntSymbol)
       << "ntSymbol to be one of the non-terminal symbols given in the "
          "predeclaration";
+  for (size_t i = 0, n = rules.size(); i < n; ++i)
+  {
+    CVC4_API_ARG_AT_INDEX_CHECK_EXPECTED(
+        !containsFreeVariables(rules[i]), rules[i], rules, i)
+        << "a constant term or a term containing synthFun/synthInv parameters "
+           "and non-terminal symbols of the grammar";
```
Comment (ajreynol) [inline src/api/cvc4cpp.cpp:3688]:
```suggestion
      << "Expected a term whose free variables are limited to synthFun/synthInv parameters "
         "and non-terminal symbols of the grammar";
```

**Review by ajreynol — CHANGES_REQUESTED** [review]

File: src/api/cvc4cpp.cpp (hunk 4 above)
Comment (aniemetz) [reply src/api/cvc4cpp.cpp:3665]: @ajreynol `Expected` is not necessary, it's part of the message of the macro.

File: src/api/cvc4cpp.cpp (hunk 5 above)
Comment (aniemetz) [reply src/api/cvc4cpp.cpp:3688]: Ditto

**Review by ajreynol — APPROVED** [review]

**Review by aniemetz — APPROVED** [review]

**Pull request merged**